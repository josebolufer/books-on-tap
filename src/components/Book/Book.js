import PropTypes from 'prop-types'
import Container from '@material-ui/core/Container'
import { makeStyles } from '@material-ui/core/styles'
import Button from '@material-ui/core/Button'
import CardMedia from '@material-ui/core/CardMedia'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import Box from '@material-ui/core/Box'
import LinearProgress from '@material-ui/core/LinearProgress'

const useStyles = makeStyles((theme) => ({
  root: {
    width: '100%',
    '& > * + *': {
      marginTop: theme.spacing(2),
    },
  },
  cardGrid: {
    paddingTop: theme.spacing(8),
    paddingBottom: theme.spacing(8),
  },
}))

const Book = ({ book, addToShoppingBag, fetching }) => {
  const { author, thumbnail, title, description, price = 0, stockAmount } = book
  const { firstName = '', lastName = '' } = author || {}
  const classes = useStyles()

  const handleAddToBag = () => {
    book.quantity = 1
    addToShoppingBag(book)
  }

  if (fetching)
    return (
      <div data-testid='book-list-fetching' className={classes.root}>
        <LinearProgress color='secondary' />
      </div>
    )

  return (
    <Container className={classes.cardGrid} maxWidth='xl'>
      <Grid container spacing={3}>
        <Grid item xs={12} sm={6} md={6}>
          <CardMedia
            alt={title}
            component='img'
            src={
              thumbnail
                ? thumbnail
                : 'https://via.placeholder.com/400.png?text=Loading...'
            }
            title={title}
          />
        </Grid>
        <Grid item xs={12} sm={6} md={6}>
          <Typography variant='h4' gutterBottom color='primary'>
            <Box fontWeight='fontWeightBold' m={1}>
              {title}
            </Box>
          </Typography>
          <Typography variant='h5' gutterBottom>
            <Box fontWeight='fontWeightLight' m={1} display='inline'>
              Price:
            </Box>
            <Box fontWeight='fontWeightBold' m={1} display='inline'>
              Â£{price.toFixed(2)}
            </Box>
          </Typography>
          <Typography variant='h5' gutterBottom>
            <Box fontWeight='fontWeightLight' m={1} display='inline'>
              Author:
            </Box>
            <Box fontWeight='fontWeightBold' m={1} display='inline'>
              {`${firstName} ${lastName}`}
            </Box>
          </Typography>
          <Typography variant='body1' gutterBottom>
            {description}
          </Typography>
          <Typography variant='h5' gutterBottom>
            {stockAmount < 1 ? 'Out of stock' : `In stock: ${stockAmount}`}
          </Typography>
          <Button
            onClick={() => handleAddToBag()}
            size='large'
            fullWidth
            disabled={stockAmount < 1}
            variant='outlined'
            color='primary'
          >
            ADD TO BAG
          </Button>
        </Grid>
      </Grid>
    </Container>
  )
}

Book.propTypes = {
  book: PropTypes.object.isRequired,
  addToShoppingBag: PropTypes.func.isRequired,
  fetching: PropTypes.bool,
}

Book.defaultProps = {
  fetching: false,
}

export default Book
