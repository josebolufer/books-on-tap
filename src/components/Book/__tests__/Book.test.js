import { render, screen, fireEvent } from '@testing-library/react'
import Book from '../Book'

const mockedBook = {
  author: { firstName: 'Jose', lastName: 'Bolufer' },
  id: '1',
  description: 'The story about a boy that grew up in London',
  isbn: 'asdfghghghghg',
  price: 10,
  stockAmount: 1,
  thumbnail: 'Image 1',
  title: 'A live in London',
}

const addToShoppingBag = jest.fn()

describe('@renders', () => {
  it('in default state', () => {
    const { asFragment } = render(
      <Book book={mockedBook} addToShoppingBag={addToShoppingBag} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  describe('when book is not out of stock', () => {
    it('should render the book image', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      expect(screen.getByAltText('A live in London')).toHaveAttribute(
        'src',
        'Image 1'
      )
    })

    it('should render the price of the book', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      screen.getByText((content, node) => {
        const hasText = (node) => node.textContent === 'Price:£10.00'
        const nodeHasText = hasText(node)
        const childrenDontHaveText = Array.from(node.children).every(
          (child) => !hasText(child)
        )

        return nodeHasText && childrenDontHaveText
      })
    })

    it('should render the firstname and lastname of the author of the book', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      screen.getByText((content, node) => {
        const hasText = (node) => node.textContent === 'Author:Jose Bolufer'
        const nodeHasText = hasText(node)
        const childrenDontHaveText = Array.from(node.children).every(
          (child) => !hasText(child)
        )

        return nodeHasText && childrenDontHaveText
      })
    })

    it('should render the title of the book', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      expect(
        screen.getByText('The story about a boy that grew up in London')
      ).toBeInTheDocument()
    })

    it('should render the number of books in stock', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      expect(screen.getByText('In stock: 1')).toBeInTheDocument()
    })

    it('should render the `add to bag` not being dissabled', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      expect(
        screen.getByText('ADD TO BAG').closest('button')
      ).not.toBeDisabled()
    })

    it('should dispatch `addToShoppingBag` action when clicking on `ADD TO BAG`', () => {
      render(<Book book={mockedBook} addToShoppingBag={addToShoppingBag} />)

      fireEvent.click(screen.getByText('ADD TO BAG'))

      expect(addToShoppingBag).toHaveBeenCalledTimes(1)
      expect(addToShoppingBag).toHaveBeenCalledWith(mockedBook)
    })
  })

  describe('when book is out of stock', () => {
    it('should render the book image', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      expect(screen.getByAltText('A live in London')).toHaveAttribute(
        'src',
        'Image 1'
      )
    })

    it('should render the price of the book', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      screen.getByText((content, node) => {
        const hasText = (node) => node.textContent === 'Price:£10.00'
        const nodeHasText = hasText(node)
        const childrenDontHaveText = Array.from(node.children).every(
          (child) => !hasText(child)
        )

        return nodeHasText && childrenDontHaveText
      })
    })

    it('should render the firstname and lastname of the author of the book', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      screen.getByText((content, node) => {
        const hasText = (node) => node.textContent === 'Author:Jose Bolufer'
        const nodeHasText = hasText(node)
        const childrenDontHaveText = Array.from(node.children).every(
          (child) => !hasText(child)
        )

        return nodeHasText && childrenDontHaveText
      })
    })

    it('should render the title of the book', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      expect(
        screen.getByText('The story about a boy that grew up in London')
      ).toBeInTheDocument()
    })

    it('should render the out of stock text', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      expect(screen.getByText('Out of stock')).toBeInTheDocument()
    })

    it('should render the `add to bag` being dissabled', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      expect(screen.getByText('ADD TO BAG').closest('button')).toBeDisabled()
    })

    it('should not dispatch `addToShoppingBag` action when clicking on `ADD TO BAG`', () => {
      render(
        <Book
          book={{ ...mockedBook, stockAmount: 0 }}
          addToShoppingBag={addToShoppingBag}
        />
      )

      fireEvent.click(screen.getByText('ADD TO BAG'))

      expect(addToShoppingBag).toHaveBeenCalledTimes(0)
    })
  })
})
