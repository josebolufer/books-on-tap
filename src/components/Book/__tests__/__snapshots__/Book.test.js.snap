// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@renders in default state 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-cardGrid-2 MuiContainer-maxWidthXl"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6"
      >
        <img
          alt="A live in London"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
          src="Image 1"
          title="A live in London"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-colorPrimary MuiTypography-gutterBottom"
        >
          <div
            class="MuiBox-root MuiBox-root-3"
          >
            A live in London
          </div>
        </h4>
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          <div
            class="MuiBox-root MuiBox-root-4"
          >
            Price:
          </div>
          <div
            class="MuiBox-root MuiBox-root-5"
          >
            Â£10.00
          </div>
        </h5>
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          <div
            class="MuiBox-root MuiBox-root-6"
          >
            Author:
          </div>
          <div
            class="MuiBox-root MuiBox-root-7"
          >
            Jose Bolufer
          </div>
        </h5>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom"
        >
          The story about a boy that grew up in London
        </p>
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          In stock: 1
        </h5>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-outlinedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            ADD TO BAG
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
